### DISCUSSION TITLE
Standardize errors for Directus and extensions by exporting from @api/errors

### DISCUSSION THREAD

**Discussion by Zehir** [directus/directus#19872] [Feature Requests]
### Summary

Currently `isDirectusError()` method accept a generic argument for the kind of errors but there is no way to import the extension from the errors source.

It's will also provide the extension creator to throw errors with the same format of the directus ones. Like if an auth extension need to throw an `MethodNotAllowedError` error.

It's a way to standardise errors for extensions and not having errors that is almost identical but not excatly the same.

### Basic Example

The ability to import api errors to be used in extension

With current namming of the const in `@directus/api` if they are exported we could do :
```typescript
import { ErrorCode } from '@directus/api'
import type { RecordNotUniqueErrorExtensions } from '@directus/api'

if (isDirectusError<RecordNotUniqueErrorExtensions>(error, ErrorCode.RecordNotUnique)) {
  switch (error.extensions.collection) {
    case 'bundles':
      switch (error.extensions.field) {
        case 'id':
          return 'Bundle with same id already exists'
  }
}
```

But maybe it's worth reworking isDirectusError to accept a `createError` return instead of a string;

```typescript
import { RecordNotUnique } from '@directus/errors'
// RecordNotUnique is createError<RecordNotUniqueErrorExtensions>(...)

if (isDirectusError(error, RecordNotUnique)) {
  switch (error.extensions.collection) {
    case 'bundles':
      switch (error.extensions.field) {
        case 'id':
          return 'Bundle with same id already exists'
  }
}
```

### Motivation

I am using a custom endpoint to insert data into the database. It's an uploaded "zip" file with many data inside that need to be splited up to multiple tables with some extended validation. So for that case I was not able to use the included validator.

So I need to insert data into multiple collections with some uniques constraint so I may get errors from multiple sources.

And to give clear errors message I need to read from it to determine what to do.

### Detailed Design

The `@directus/errors` package will need to have any errors used by directus from the `createError()` call. Any errors that is used should be migrated to that package to be referenced from the api package.

I think it's worth exporting both string codes like the `ErrorCode` enum here :
https://github.com/directus/directus/blob/01f0a1814709d792cc15d8d4542e3ca16c7848db/api/src/errors/codes.ts#L1-L29

Migrate any `createError` that could be without side effects in the `@directus/errors` package.
For example :
https://github.com/directus/directus/blob/01f0a1814709d792cc15d8d4542e3ca16c7848db/api/src/errors/record-not-unique.ts#L25-L29

### Must Have:
- Exported errors codes
- Exported errors extension types

### Could Have:
- Exported errors created with `createError`
- Refactor `isDirectusError` to accept `createError` response
- Limit `createError` to **not** accept codes that are defined in the well-known errors.

### Drawbacks

If we limit `createError` code param to exclude any codes defined in the well-known codes, users will get a typescript errors, they will need to use the exported one instead of making a new one.

### Alternatives

As an alternative I was able to make it work with :

```typescript
import type { RecordNotUniqueErrorExtensions } from '@directus/api/dist/errors/record-not-unique'

const result = isDirectusError<RecordNotUniqueErrorExtensions>(error, 'RECORD_NOT_UNIQUE')
```

And in tsconfig.json
```json
    "include": [
        "./src/**/*.ts",
        "@directus/api/dist/errors/record-not-unique.d.ts"
    ]
```

But not really a good idea to do that.

### Adoption Strategy

No kreaking change as you just add exports and you can check `typeof code` for the `isDirectusError` method

**Comment by rijkvanzanten** [comment]
I'm a little hesitant to expose the ones directly exported from `/api`. I think I'd prefer moving the system-used ones into `/errors`, so you can just import them directly from there (and keep all the error stuff nice and in the same scope 🙂 )

**Reply by Zehir** [reply]
Yes that was I was thinking too, any errors in the errors package and imported back when needed like in the api package.

> Migrate any createError that could be without side effects in the @directus/errors package

**Reply by rijkvanzanten** [reply]
@paescuj Any hesitations with doing this? Feels like a small codebase optimization rather than a big new feature to me. I'm down to just refactor this through and get it done if we can't think of any good reasons why not to do this 🙂

**Reply by paescuj** [reply]
Nope, I don't see an issue with this! Sounds like a good plan 👍

**Comment by rijkvanzanten** [comment]
This feature request has been accepted and is queued to be implemented! You can follow along with the progress here: https://github.com/directus/directus/issues/19906

**Comment by rijkvanzanten** [comment]
This was completed!
